include/envoy/network/io_handle.h

```diff
@@ -7,27 +7,93 @@
 namespace Envoy {
 namespace Network {
 
+namespace Address {
+class Instance;
+} // namespace Address
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementation.
+ */
+template <typename T> struct IoHandleCallResult {
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  int errno_;
+};
+
+typedef IoHandleCallResult<int> IoHandleCallIntResult;
+typedef IoHandleCallResult<ssize_t> IoHandleCallSizeResult;
+
 /**
  * IoHandle: an abstract interface for all I/O operations
  */
 class IoHandle {
 public:
-  IoHandle() {}
-
   virtual ~IoHandle() {}
 
   /**
    * Return data associated with IoHandle.
    *
    * TODO(sbelair2) remove fd() method
+   * We probably still need some method similar to this one for
+   * evconnlistener_new(). Or We can move it to IoSocketHandle and down cast the
+   * IoHandle to IoSocketHandle wherever needed.
    */
```

Comment:
My opinion is that it would be cleaner to move this to the derived IoSocketHandle- what does the fd() method resolve to for interfaces that are not sockets? Either that, or there can be a generic method to manage a descriptor of sorts, such that for sockets,  the descriptor is really just the fd. And for other interfaces it might be a session id, etc.

Reply:
@danzh1989 @sbelair2 I don't think I'm grokking this convo fully. Can one or both of you expand? I don't think we should still need this method so I'm likely missing something.

Reply:
@mattklein123 Maybe I missed some earlier conversation, but what shall be passed into evconnlistener_new() if there is no fd() or something similar in interface?

Reply:
We could probably figure out an interface strategy in which the socket actually creates the listener somehow, but for now, I would probably just do a dynamic_cast. It's a little ugly, but VPP or similar will have to have their own listener implementation, so it should fit together OK? Thoughts?

Reply:
@mattklein123 Do you mean to make fd() a method to IoSocketHandle only and dynamic_cast the IoHhandle object when we create listener? That sounds reasonable to me.

Reply:
Yes that's what I meant. IMO that's OK for now.